### DISCUSSION TITLE
sqlite: Cant run migrate:latest - Error FOREIGN KEY constraint failed

### DISCUSSION THREAD

**Discussion by scolastico** [directus/directus#15020] [Archive]
### Describe the Bug

If i run `npx directus database migrate:latest` or try to create new tables with relations in it in directus i get an "FOREIGN KEY constraint failed" error.

### To Reproduce

run `npx directus database migrate:latest`

### Errors Shown

```log
12:37:40 ✨ Running migrations...
12:37:40 🔍 Enabling SQLite Foreign Keys support...
12:37:40 🔍 [0.680ms] select * from `directus_migrations` order by `version` asc []
12:37:40 ✨ Applying Update Notifications Timestamp Column...
12:37:40 🔍 [0.122ms] SELECT type, sql FROM sqlite_master WHERE (type='table' OR (type='index' AND sql IS NOT NULL)) AND tbl_name='directus_notifications' []
12:37:40 🔍 [0.075ms] PRAGMA foreign_keys []
12:37:40 🔍 [0.064ms] PRAGMA foreign_keys = OFF []
12:37:40 🔍 [0.265ms] CREATE TABLE `_knex_temp_alter472` (`id` integer PRIMARY KEY AUTOINCREMENT NOT NULL, `timestamp` datetime DEFAULT CURRENT_TIMESTAMP, `status` varchar(255) DEFAULT 'inbox', `recipient` char(36) NOT NULL, `sender` char(36), `subject` varchar(255) NOT NULL, `message` text, `collection` varchar(64), `item` varchar(255), FOREIGN KEY (`recipient`) REFERENCES `directus_users` (`id`) ON DELETE CASCADE, FOREIGN KEY (`sender`) REFERENCES `directus_users` (`id`)) []
12:37:40 🔍 [0.174ms] INSERT INTO "_knex_temp_alter472" SELECT * FROM "directus_notifications"; []
12:37:40 🔍 [0.176ms] DROP TABLE "directus_notifications" []
12:37:40 🔍 [1.370ms] ALTER TABLE "_knex_temp_alter472" RENAME TO "directus_notifications" []
12:37:40 🔍 [0.199ms] PRAGMA foreign_key_check []
12:37:40 🔍 [0.061ms] PRAGMA foreign_keys = ON []
12:37:40 🚨 FOREIGN KEY constraint failed 
Error: FOREIGN KEY constraint failed
    at client.transaction.connection (/mnt/data/workspace/***/directus/node_modules/.pnpm/knex@2.2.0_wm42kvundpz33c2ast42l7zyba/node_modules/knex/lib/execution/runner.js:254:23)
```

### What version of Directus are you using?

9.15.1

### What version of Node.js are you using?

v16.16.0

### What database are you using?

SQLite

### What browser are you using?

n/a

### How are you deploying Directus?

running locally

**Comment by barakplasma** [comment]
got the same error when migrating from 9.14.5 to 9.15.1 (Node 18.7.0 on Ubuntu 22.04)

once when switching the docker image version on the same sqlite db within docker-compose
```
directus                 | 11:34:51 ✨ Loaded extensions: platemapping
directus                 | 11:34:52 ✨ Initializing bootstrap...
directus                 | 11:34:52 ✨ Database already initialized, skipping install
directus                 | 11:34:52 ✨ Running migrations...
directus                 | 11:34:52 ✨ Applying Update Notifications Timestamp Column...
directus                 | Error: FOREIGN KEY constraint failed
directus                 |     at client.transaction.connection (/directus/node_modules/knex/lib/execution/runner.js:254:23)
```

and once with the same command as scolastico 
```
michael@quris-michael:~/Projects/quris/data-pipeline/quris$ npx directus database migrate:latest
14:38:32 ✨ Loaded extensions: platemapping
14:38:32 ✨ Running migrations...
14:38:32 ✨ Applying Update Notifications Timestamp Column...
14:38:32 🚨 FOREIGN KEY constraint failed 
Error: FOREIGN KEY constraint failed
    at client.transaction.connection (/home/michael/Projects/quris/data-pipeline/quris/node_modules/knex/lib/execution/runner.js:254:23)
```

**Comment by barakplasma** [comment]
I believe it's related to issue #14100 and PR https://github.com/directus/directus/pull/14323 / this migration:  https://github.com/directus/directus/blob/ed82a01c2c93eb74df91385b554692d2e3ea3caa/api/src/database/migrations/20220801A-update-notifications-timestamp-column.ts#L7-L10 which sqlite doesn't support due to it's lack of ALTER TABLE support to make a column nullable. (see [sqlite ALTER TABLE](https://www.sqlite.org/draft/lang_altertable.html#:~:text=SQLite%20supports%20a%20limited%20subset%20of%20ALTER%20TABLE.%20The%20ALTER%20TABLE%20command%20in%20SQLite%20allows%20these%20alterations%20of%20an%20existing%20table%3A%20it%20can%20be%20renamed%3B%20a%20column%20can%20be%20renamed%3B%20a%20column%20can%20be%20added%20to%20it%3B%20or%20a%20column%20can%20be%20dropped%20from%20it.) docs for the limited set of alter table it does support)

I don't know of a Knex level work around for this. When I needed to change a column, I did it by adding a new column, inserting the old column to the new column, and then dropping the old column, and renaming the new column.

**Comment by azrikahar** [comment]
@barakplasma It is very likely caused by that PR, but IIRC the "make a column nullable" change was also applicable here when Flows was released:

https://github.com/directus/directus/blob/ed82a01c2c93eb74df91385b554692d2e3ea3caa/api/src/database/migrations/20220429D-drop-non-null-from-sender-of-notifications.ts#L4-L6

so assuming that migration worked on the same `directus_notifications` table, it's odd how this migration is causing error. 

Just to confirm, is there existing notifications in that table before you ran into this migration?

**Comment by barakplasma** [comment]
I don't have any notifications in my sqlite db.
```sql
sqlite> .schema directus_notifications 
CREATE TABLE IF NOT EXISTS "directus_notifications" (`id` integer PRIMARY KEY AUTOINCREMENT NOT NULL, `timestamp` datetime NOT NULL, `status` varchar(255) DEFAULT 'inbox', `recipient` char(36) NOT NULL, `sender` char(36), `subject` varchar(255) NOT NULL, `message` text, `collection` varchar(64), `item` varchar(255), FOREIGN KEY (`recipient`) REFERENCES `directus_users` (`id`) ON DELETE CASCADE, FOREIGN KEY (`sender`) REFERENCES `directus_users` (`id`));
sqlite> select count(id) from directus_notifications;
0
```

**Comment by scolastico** [comment]
I neither. Table is there but empty.

**Comment by azrikahar** [comment]
Unfortunately I can't seem to reproduce this. Is there any other non out-of-the-box configs that may be at play here? perhaps such as custom extensions, custom fields on `directus_notifications`, etc? 🤔

Example screenshot:

![WindowsTerminal_2xKaX2gfoT](https://user-images.githubusercontent.com/42867097/183805235-c8214068-6e44-43e2-9336-4209df6a33ae.png)

A step by step list of what I did in the screenshot (step 1 and 2 is out of frame):

1. Create new manual project, with a copy of standard env file (what we typically get when using `npm init directus-project <project-name>`).
2. Install `directus@9.14.5`.
3. Run `npx directus bootstrap`.
4. Run `npm list directus` to confirm it's `9.14.5`.
5. Run `npm i directus@9.15.1` to install the newer version.
6. Run `npm list directus` to confirm it's `9.15.1`.
7. Add `LOG_LEVEL=trace` in env file to see more info later.
8. Run `npx directus database migrate:latest` to run newer migrations.

Just out of curiosity, I've also tried repeating step 1 to 4, then creating a new notification, and follow through step 5 to 8 without any issue.

**Comment by scolastico** [comment]
If you leave me a email adress here i can send you the sqlite file i'm using. Also fyi i downgraded to 9.14.1 and everything works like a charm again.

**Comment by azrikahar** [comment]
> If you leave me a email adress here i can send you the sqlite file i'm using.

Sure thing, feel free to send it to `azri@directus.io` 🙂

> Also fyi i downgraded to 9.14.1 and everything works like a charm again.

Does this mean you also ran `npx directus database migrate:down` to undo it before downgrading? I believe the issue should only be isolated with the Update Notifications Timestamp column migration, and `9.15.1` should've still ran without any issue, but the failed migration is currently a blocker in your case. If you can also try manually entering the sqlite file and changing `directus_notifications`'s `timestamp` column to nullable and try to upgrade again, that'd be much appreciated!

**Comment by scolastico** [comment]
> Does this mean you also ran `npx directus database migrate:down` to undo it before downgrading? I believe the issue should only be isolated with the Update Notifications Timestamp column migration, and `9.15.1` should've still ran without any issue, but the failed migration is currently a blocker in your case. If you can also try manually entering the sqlite file and changing `directus_notifications`'s `timestamp` column to nullable and try to upgrade again, that'd be much appreciated!

No i simply make backups before updates and reverted to that state. No backups = no compassion.

Also i have send you the file. :)

**Comment by azrikahar** [comment]
@scolastico thanks again for the file! Turns out to be a rather odd case here.

The `FOREIGN KEY constraint failed` error seems to be due to `directus_files` and `directus_users`. This is the code in Knex (where this error is thrown):

```js
if (foreignViolations.length > 0) {
  throw new Error('FOREIGN KEY constraint failed');
}
```

If we check `foreignViolations` triggeredd with your file, we'll get:

```js
{
  foreignViolations: [
    {
      table: 'directus_files',
      rowid: 1,
      parent: 'directus_users',
      fkid: 2
    },
    {
      table: 'directus_files',
      rowid: 2,
      parent: 'directus_users',
      fkid: 2
    },
    {
      table: 'directus_files',
      rowid: 3,
      parent: 'directus_users',
      fkid: 2
    },
    {
      table: 'directus_files',
      rowid: 4,
      parent: 'directus_users',
      fkid: 2
    }
  ]
}
```

Based on that info and checking your `directus_files` table, it does seem like the entries 1 to 4 are having empty (but not NULL) `uploaded_by` values:

![DB_Browser_for_SQLite_wVZfuTJPUK](https://user-images.githubusercontent.com/42867097/184091188-66277620-44bc-44bf-b939-f9362d6438bf.png)

After turning these 4 values into NULL properly:

![DB_Browser_for_SQLite_FFzMzM6257](https://user-images.githubusercontent.com/42867097/184091358-10758d30-27b7-4cbd-8974-960f59a3d91a.png)

I am able to run the "Update Notifications Timestamp Column" migration without any issue now. Would you mind trying this out and let me know if that's the case?

If this _does_ resolve it, I do wonder if the non NULL `uploaded_by` is from unintentional manual editings, or something caused by Directus. However I'm inclined to believe it's not Directus due to #9507, so it'd be odd that it can be an empty non NULL value without modifying the constraint or values. 🤔

**Comment by barakplasma** [comment]
I can't recreate going from 9.14.5 to 9.15.1 using a db like chinook (see my attempt to recreate repo at  https://github.com/barakplasma/directus-issue-14976 ) . I also tried uploading a file before migrating.

I started from directus@9.12 in the original DB if that makes a difference for recreating. I can try sending my DB to you, if it helps.

**Comment by azrikahar** [comment]
> I can't recreate going from 9.14.5 to 9.15.1 using a db like chinook (see my attempt to recreate repo at https://github.com/barakplasma/directus-issue-14976 ) . I also tried uploading a file before migrating.
> 
> I started from directus@9.12 in the original DB if that makes a difference for recreating. I can try sending my DB to you, if it helps.

@barakplasma Granted I'm not familiar with Chinook, but I was able to run `directus bootstrap` on the linked sqlite file (https://github.com/lerocha/chinook-database/blob/e7e6d5f008e35d3f89d8b8a4f8d38e3bfa7e34bd/ChinookDatabase/DataSources/Chinook_Sqlite.sqlite) in your repo without any issue.

Can you try inspecting `foreignViolations` in knex like I shared above? Adding a simple `console.log(foreignViolations)` before that if condition should suffice. Alternatively you can send the db to same email.

EDIT: I've also tried https://github.com/barakplasma/directus-issue-14976/blob/using-database-migrate/Chinook_Sqlite.sqlite, but I presume this is already migrated? your logs seems to indicate it's successful 🤔

**Comment by scolastico** [comment]
I checked the git logs and it seems that these null's came in during the version change from 19.13.0 -> 19.14.1 (several things were done with the commit so it's possible that it's not from the version change). I also noticed that further to the right in the table fields like `uploaded_on` were also set to 0.

I changed some things in the tables, but I'm (99%) sure I didn't touch that. My guess, if this issue is a one timer, would be that something corrupted the file and Directus or my ide fixed it, but the bytes were missing or something idk? ¯\_(ツ)_/¯

Anyway after fixing the db the migration works like a charm.

Thanks for your help ^^

**Comment by azrikahar** [comment]
@scolastico thanks for confirming! Yeah unfortunately it's quite hard to tell what happened, but if you ever find out the incorrect data stems somewhere from Directus, do let us know 👍

@barakplasma seems like this is a similar issue with OP, but in your case, it's not related to directus tables at all 🤔

These are what's triggering the foreignViolations:

```js
{
  foreignViolations: [
    { table: 'platereadersample', rowid: 1, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 2, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 3, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 4, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 5, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 6, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 7, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 8, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 9, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 10, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 11, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 12, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 13, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 14, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 15, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 16, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 17, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 18, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 19, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 20, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 21, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 22, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 23, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 24, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 25, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 26, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 27, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 28, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 29, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 30, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 31, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 32, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 33, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 34, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 35, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 37, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 38, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 39, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 40, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 41, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 42, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 43, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 44, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 45, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 46, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 47, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 48, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 49, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 50, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 51, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 52, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 53, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 54, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 55, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 56, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 57, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 58, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 59, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 60, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 61, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 62, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 63, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 64, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 65, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 66, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 67, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 68, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 69, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 70, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 71, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 72, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 73, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 74, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 75, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 76, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 77, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 78, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 79, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 80, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 81, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 82, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 83, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 84, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 85, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 86, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 87, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 88, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 89, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 90, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 91, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 92, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 93, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 94, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 95, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 96, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 97, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 98, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 99, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 100, parent: 'file', fkid: 0 },
    { table: 'platereadersample', rowid: 101, parent: 'file', fkid: 0 },
    ... 2011 more items
  ]
}
```

When we check in `files` table, there are only 4 files with IDs `24`, `94`, `95`, and `113`:

![DB_Browser_for_SQLite_jen4LpsAle](https://user-images.githubusercontent.com/42867097/184104194-34e6bafb-4e6f-472c-8efb-409be091bc7a.png)

But when we check `platreadersample` table, there are many `file` foreign key that does not exist anymore:

![DB_Browser_for_SQLite_Uh78JuB0er](https://user-images.githubusercontent.com/42867097/184104296-d2e92fcd-2140-40e7-9bf6-eb18d04f713f.png)

You will need to resolve these manually for now, as I notice that your `file` column is set as NOT NULL, so we can't just turn them to null. I believe once you resolve this, the migration will work (once again turns out the migration will still get blocked even when it has nothing to do with the `files` and `platereadersample`. I'm not familiar with this interaction, but understandably it's trying to reconcile all issues during the process).

I'll move this to a discussion for now since there doesn't seem to be a shared table with issue in both cases. However I also acknowledge that these 2 cases has the similarity of foreign keys that _should not_ exist, ended up existing anyway. Hopefully we can find out the actual root cause 👍

**Reply by barakplasma** [reply]
Thanks @azrikahar !

A nice DX improvement would be if log level trace would surface these error logs from knex

**Reply by azrikahar** [reply]
Definitely agree, but this error is technically "swalllowed" by knex itself and I had to tinker with the knex package, so Directus can't really surface it, at least from my understanding 🤔

**Reply by barakplasma** [reply]
Yeah, I don't see a way for Directus to surface that info from Knex either :disappointed: 
I made a PR https://github.com/knex/knex/pull/5300 on Knex to add that ability via environment variable DEBUG=knex:*

**Comment by jbogdani** [comment]
Hello all, 
I have a similar issue updating from v. `9.24` to v. `9.25` (and up). I'm using the docker version and I'm getting the following error:
```
Error: FOREIGN KEY constraint failed
    at client.transaction.connection (/directus/node_modules/.pnpm/knex@2.4.2_mysql@2.18.1_pg@8.10.0_sqlite3@5.1.6_tedious@15.1.3/node_modules/knex/lib/execution/runner.js:254:23)
INFO: Initializing bootstrap...
INFO: Database already initialized, skipping install
INFO: Running migrations...
INFO: Applying Update Material Icons...
exited with code 1
```

I tried to look at `directus_files` and `directus_users` but I did not fount any empty (and non NULL) values: everything seems to be OK.

Any help to debug this issue?

Thank you
